### PR TITLE
Update to manage the condition of GitHubSource webhook

### LINKED ISSUE #994: Github Source reports everything as fine, when it is not fine
I would expect the Github source to validate the event types match the rest of the configuration. 
 Ideally this would be done in the webhook, but otherwise surfacing failures in the controller would be expected.

What I actually get is as follows...

The resource reports everything is successful 🎉 
```
kubectl get githubsource knative-serving -oyaml
apiVersion: sources.knative.dev/v1alpha1
kind: GitHubSource
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"sources.knative.dev/v1alpha1","kind":"GitHubSource","metadata":{"annotations":{},"name":"knative-serving","namespace":"default"},"spec":{"accessToken":{"secretKeyRef":{"key":"accessToken","name":"github-hook-secret"}},"eventTypes":["check_suite","commit_comment","create","delete","deployment","deployment_status","fork","gollum","installation","integration_installation","issue_comment","issues","label","member","membership","milestone","organization","org_block","page_build","ping","project_card","project_column","project","public","pull_request","pull_request_review","pull_request_review_comment","push","release","repository","status","team","team_add","watch"],"ownerAndRepository":"knative/serving","secretToken":{"secretKeyRef":{"key":"secretToken","name":"github-hook-secret"}},"sink":{"ref":{"apiVersion":"eventing.knative.dev/v1alpha1","kind":"Broker","name":"default"}}}}
  creationTimestamp: "2020-03-06T19:18:01Z"
  finalizers:
  - github-source-controller
  generation: 2
  name: knative-serving
  namespace: default
  resourceVersion: "40585"
  selfLink: /apis/sources.knative.dev/v1alpha1/namespaces/default/githubsources/knative-serving
  uid: 7843c4ce-ffdc-44e2-adb2-6b151f953ee2
spec:
  accessToken:
    secretKeyRef:
      key: accessToken
      name: github-hook-secret
  eventTypes:
  - check_suite
  - commit_comment
  - create
  - delete
  - deployment
  - deployment_status
  - fork
  - gollum
  - installation
  - integration_installation
  - issue_comment
  - issues
  - label
  - member
  - membership
  - milestone
  - organization
  - org_block
  - page_build
  - ping
  - project_card
  - project_column
  - project
  - public
  - pull_request
  - pull_request_review
  - pull_request_review_comment
  - push
  - release
  - repository
  - status
  - team
  - team_add
  - watch
  ownerAndRepository: knative/serving
  secretToken:
    secretKeyRef:
      key: secretToken
      name: github-hook-secret
  sink:
    ref:
      apiVersion: eventing.knative.dev/v1alpha1
      kind: Broker
      name: default
status:
  conditions:
  - lastTransitionTime: "2020-03-06T19:18:01Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2020-03-06T19:18:01Z"
    status: "True"
    type: SecretsProvided
  - lastTransitionTime: "2020-03-06T19:21:16Z"
    status: "True"
    type: SinkProvided
  observedGeneration: 2
  sinkUri: http://default-broker.default.svc.cluster.local
```

However, there's no webhook on github (wtf!)

Here are the controller logs:
```
{"level":"error","ts":"2020-03-06T19:21:17.335Z","logger":"github-controller.github-source-controller","caller":"controller/controller.go:409","msg":"Reconcile error","knative.dev/controller":"github-source-controller","error":"failed to create webhook: failed to create the webhook: POST https://api.github.com/repos/knative/serving/hooks: 422 Validation Failed [{Resource:Hook Field: Code:custom Message:These events are not allowed for this hook: installation, integration_installation, membership, organization, org_block, team}]","stacktrace":"knative.dev/eventing-contrib/vendor/knative.dev/pkg/controller.(*Impl).handleErr\n\tknative.dev/eventing-contrib/vendor/knative.dev/pkg/controller/controller.go:409\nknative.dev/eventing-contrib/vendor/knative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/eventing-contrib/vendor/knative.dev/pkg/controller/controller.go:395\nknative.dev/eventing-contrib/vendor/knative.dev/pkg/controller.(*Impl).Run.func2\n\tknative.dev/eventing-contrib/vendor/knative.dev/pkg/controller/controller.go:343"}
```

🤦‍♂ 

cc @n3wscott @grantr

### PR DESCRIPTION
**Pull request by tom24d**
Fixes #994 

## Proposed Changes

- 🧽managing the condition of GitHub webhook in GitHubSource reconciler.
- 🧽update the tests that validate the combination of conditions that include webhook.

### CODE REVIEW

File: github/pkg/apis/sources/v1alpha1/githubsource_types_test.go [hunk 1]
```diff
@@ -285,13 +329,32 @@ func TestGitHubSourceStatusGetCondition(t *testing.T) {
...
+			s.MarkSink("uri://example")
+			s.MarkSecrets()
+			s.MarkEventTypes()
+			s.MarkNoWebhook("Testing", "hi%s", "")
+			return s
+		}(),
+		condQuery: GitHubSourceConditionReady,
+		want: &apis.Condition{
+			Type:   GitHubSourceConditionReady,
+			Status: corev1.ConditionFalse,
```
Comment (mattmoor) [inline github/pkg/apis/sources/v1alpha1/githubsource_types_test.go]: Format Go code:
```suggestion
			Type:    GitHubSourceConditionReady,
			Status:  corev1.ConditionFalse,
```

**Comment by tom24d** [comment]
@googlebot I signed it!

File: github/pkg/apis/sources/v1alpha1/githubsource_types_test.go (hunk 1 above)
Comment (tom24d) [reply github/pkg/apis/sources/v1alpha1/githubsource_types_test.go]: Thanks! It has been done and pushed. Does it look good?

File: github/pkg/apis/sources/v1alpha1/githubsource_types.go [hunk 2]
```diff
@@ -130,6 +130,10 @@ const (
 	// GitHubSource has been configured with event types.
 	GitHubSourceConditionEventTypesProvided apis.ConditionType = "EventTypeProvided"
 
+	//GitHubSourceConditionWebhookProvided has a status True when the
+	//GitHubSource has been configured with a webhook.
```
Comment (mattmoor) [inline github/pkg/apis/sources/v1alpha1/githubsource_types.go]:
```suggestion
	// GitHubSourceConditionWebhookProvided has a status True when the
	// GitHubSource has been configured with a webhook.
```

File: github/pkg/apis/sources/v1alpha1/githubsource_types.go [hunk 3]
```diff
@@ -130,6 +130,10 @@ const (
...
+	GitHubSourceConditionWebhookProvided apis.ConditionType = "WebhookProvided"
```
Comment (mattmoor) [inline github/pkg/apis/sources/v1alpha1/githubsource_types.go]: I think I'd use `Configured` vs. `Provided` everywhere.  `Provided` makes it seem like it's on the user to _provide_ it, and I could see that leading to confusion.

File: github/pkg/apis/sources/v1alpha1/githubsource_types.go [hunk 4]
```diff
@@ -232,6 +237,16 @@ func (s *GitHubSourceStatus) MarkNoEventTypes(reason, messageFormat string, mess
 	gitHubSourceCondSet.Manage(s).MarkFalse(GitHubSourceConditionEventTypesProvided, reason, messageFormat, messageA...)
 }
 
+// MarkWebhook sets the condition that the source has set its webhook configured.
+func (s *GitHubSourceStatus) MarkWebhook() {
```
Comment (mattmoor) [inline github/pkg/apis/sources/v1alpha1/githubsource_types.go]: `MarkWebhookConfigured`

File: github/pkg/apis/sources/v1alpha1/githubsource_types.go [hunk 5]
```diff
@@ -232,6 +237,16 @@ func (s *GitHubSourceStatus) MarkNoEventTypes(reason, messageFormat string, mess
...
+	gitHubSourceCondSet.Manage(s).MarkTrue(GitHubSourceConditionWebhookProvided)
+}
+
+// MarkNoWebhook sets the condition that the source does not have its webhook configured.
+func (s *GitHubSourceStatus) MarkNoWebhook(reason, messageFormat string, messageA ...interface{}) {
```
Comment (mattmoor) [inline github/pkg/apis/sources/v1alpha1/githubsource_types.go]: `MarkWebhookNotConfigured`

File: github/pkg/reconciler/githubsource.go [hunk 6]
```diff
@@ -160,19 +160,21 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, source *sourcesv1alpha1.
...
 			hookID, err := r.createWebhook(ctx, args)
 			if err != nil {
+				source.Status.MarkNoWebhook("WebhookCreationFailed", "%s", err)
 				return err
 			}
 			source.Status.WebhookIDKey = hookID
 		} else {
 			err := r.reconcileWebhook(ctx, args, source.Status.WebhookIDKey)
 			if err != nil {
+				source.Status.MarkNoWebhook("WebhookCreationFailed", "%s", err)
```
Comment (mattmoor) [inline github/pkg/reconciler/githubsource.go]:
```suggestion
				source.Status.MarkNoWebhook("ReconciliationFailed", err.Error())
```

File: github/pkg/reconciler/githubsource.go [hunk 7]
```diff
@@ -160,19 +160,21 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, source *sourcesv1alpha1.
 
 		// source.Status.MarkServiceDeployed(ra)
 		// TODO: Mark Deployed for the ksvc
-		// TODO: Mark some condition for the webhook status?
 		if source.Status.WebhookIDKey == "" {
 			hookID, err := r.createWebhook(ctx, args)
 			if err != nil {
+				source.Status.MarkNoWebhook("WebhookCreationFailed", "%s", err)
```
Comment (mattmoor) [inline github/pkg/reconciler/githubsource.go]: nit: the "Webhook" here feels redundant with the Type.
```suggestion
				source.Status.MarkNoWebhook("CreationFailed", err.Error())
```

**Comment by tom24d** [comment]
> Mostly nits, but this looks like the right change overall. thanks!

Thanks for the overall :)

**Comment by tom24d** [comment]
Was it right action?

**Comment by n3wscott** [comment]
/lgtm
/approve

THANKS!!!

**Pull request merged**